### ISSUE TITLE
Port to Gtk3/PyGobject/GST 1.x

### ISSUE THREAD

**Issue by thiblahute** [exaile/exaile#3]
Start from: https://bugs.launchpad.net/exaile/+bug/846379

Some blockers I can think of for PyGI/GTK3:
- [X] I don't want to _not_ have a way to select output devices, and GST 1.x got rid of the property probe interface. See https://bugzilla.gnome.org/show_bug.cgi?id=678402 (which is marked as resolved, but it's not clear to me whether it is or not)
  -> FIXED in Gst 1.4
- [ ] For Windows and OSX (both of which I use, and hope to continue to support), we currently rely on the GStreamer SDK for python bindings to GTK. As far as I can tell, it doesn't support GTK3 or GST 1, and still has the old PyGTK bindings on it. I could be wrong though.
- [ ] Right now, I don't care enough about this to do it, and besides Johannes there isn't anyone else actively doing work on the project. Will probably want to push this off as long as we can, until more people join the project.

**Comment by Vincent-C** [comment]
This is somewhat tangential to the topic of porting to gtk3 (let me know if I should open a new issue instead), but Debian is [transitioning away from gstreamer 0.10](https://lists.debian.org/debian-devel/2015/05/msg00335.html), with Ubuntu likely to follow, which means that the python bindings for gst 0.10 are likely to get removed at the same time, and thus make exaile uninstallable in Debian/Ubuntu. Is there perhaps a work in progress branch that I can cherrypick patches off from for Debian, for gst 1.0 support?

**Comment by virtuald** [comment]
@Vincent-C I believe if we move to gst 1.x, we would need to also move to GTK3 at the same time (since they both use GI, right?). This appears to be a large undertaking.

Moving to GST 1.x means we would lose cross-platform compatibility. Currently, for cross-platform distribution we rely on the old GStreamer SDK, which bundles python bindings + GTK along with GTK in a single distribution. The 1.x GStreamer releases only include binaries, and no GTK, and no python bindings. As I use Exaile on multiple platforms, I would really dislike losing support for other platforms.

I could be wrong, but I don't believe we have the manpower/will to migrate right now (unless someone speaks up! :) ). I'm certainly not motivated to do it quite yet. Furthermore, I don't believe anyone is actively working on gst 1.x or GTK3 support at this time, and that the only existing branches/etc are listed on this bug.

#75 and #76 are good motivators to move to GST 1.x.

**Comment by virtuald** [comment]
FWIW, I would be ok with getting rid of OSX support for now. I would not be ok with getting rid of Windows support, but I would also hate to lose distribution in Ubuntu/Debian even more.

**Comment by sjohannes** [comment]
There is apparently a [PyGI all-in-one installer](http://sourceforge.net/projects/pygobjectwin32/files/) which we can try.

The `gi` branch, while works to a limited degree, is still using gst 0.10. In the past I have tried converting it to gst 1.0 but didn't get far.

**Comment by virtuald** [comment]
@sjohannes ha, I had forgotten about that installer. That could be a path forward.

There's some other refactoring that I think would be good too when we go forward with this. One big one that might help remove a *lot* of boilerplate code is migrating to `guiutil.initialize_from_xml` (or something similar) instead of writing all the GtkBuilder hookup stuff manually.

**Comment by mathbr** [comment]
There's [something exciting in GTK](https://bugzilla.gnome.org/show_bug.cgi?id=612036) with that regard which makes it even easier to merge code and GtkBuilder XML resources. :-)

**Comment by virtuald** [comment]
@mathbr looks very similar in spirit to my `initialize_from_xml`, glad I'm not the only one who wanted to solve that problem. I wonder if it's supported in the python bindings yet, only saw the post about vala.

Upgrading to use something like that though would be a huge win, we have a lot of UI boilerplate in Exaile.

**Comment by mathbr** [comment]
Looks good:

```
In [1]: from gi.repository import Gtk

In [2]: Gtk.Widget.set_template
Out[2]: <bound method GObjectMeta.set_template of <class 'gi.overrides.Gtk.Widget'>>
```

Gotta love GI. :-)

**Comment by virtuald** [comment]
Cool, though presumably you would want to wrap that with an abstraction that gets rid of even more boilerplate... but now we're really offtopic.

**Comment by virtuald** [comment]
Another thought here: if we're going to migrate to GST 1.x, I'd like to remove the 'unified' and 'normal' engines and just use one or the other. I realize there were problems with the unified one, but having only one (ideally with all the features?) would make it easier to do the port.

**Comment by sjohannes** [comment]
The `gi` branch now uses gst 1.0, though only the Unified engine works. The UI is still broken in horrible ways and it segfaults occasionally but hey, progress.

By the way, I've tested that we can't mix gst 1.0 (through pygi) with gtk 2 (through pygtk); we'll have to completely migrate to the GI stack.

**Comment by virtuald** [comment]
Here's the current GTK3/GST versions available on the following distros that I have access to:

Ubuntu 14.04:
* GTK 3.10.8, GST 1.2.3

Ubuntu 15.04:
* GTK 3.14.12, GST 1.4.5

Fedora 21:
* GTK 3.14.11, GST 1.4.5

The Windows all-in-one thing mentioned above:
* GTK 3.14, GST 1.4.5

I suspect that we're probably better off at the moment not worrying about this until users start using it, but it's good to keep in mind anyways.

**Comment by genodeftest** [comment]
Fedora 22 (to be released in ~2 weeks):
* GTK 3.16.3, GST 1.4.5

**Comment by virtuald** [comment]
Alright. Given our successes so far, I think we should move forward on this and see where it goes. I've created a 'GTK3/GST1.x' label to use to track tasks/issues as we find them. I've got time to work on this during this month.. maybe some time in June, and then the rest of the summer will probably be too busy to do anything other than testing.

I've been pushing changes to the 'gi' branch too, and once we're done and the bugs are ironed out that'll be Exaile 3.5 I think (or we could do a 4.0..). Anyone interested in helping I would love to see pull requests to the gi branch also.

**Comment by virtuald** [comment]
We'll be tracking this initially using this wiki page: https://github.com/exaile/exaile/wiki/PyGI-migration-tasklist-%28GTK3-GStreamer-1.x%29 . Once it's more stable and ready for user testing we can switch to using github issues.

**Comment by virtuald** [comment]
@mathbr I implemented the composite widget thing for PyGI since I couldn't find an existing implementation: https://github.com/virtuald/pygi-composite-templates

I'll add it to xl.externals until it gets adopted by PyGI.

**Comment by virtuald** [comment]
.. and I see they have something coming. Woo.

**Comment by virtuald** [comment]
@thiblahute I took pieces of your branch and integrated it into the gi branch. Good stuff, audio is working for me now too using the unified engine.

Still a lot of weirdness happening though.

**Comment by thiblahute** [comment]
@virtuald please take whatever you want from there

I am sorry I did not have/take the time to clean that all up.

**Comment by mathbr** [comment]
@virtuald Looks interesting and just like the "syntactical sugar" mentioned in one of the first posts. ;-)

**Comment by genodeftest** [comment]
I am running into segmentation violations (SEGV) when running exaile from the gi branch. This sometimes happens when moving the mouse over menu items. I captured a backtrace. Should I report it here (or in a new bug report) or wait until exaile 3.5. becomes a testing release?

**Comment by virtuald** [comment]
Please do either of the following:

* Add it to the list of bugs on the [Wiki page](https://github.com/exaile/exaile/wiki/PyGI-migration-tasklist-%28GTK3-GStreamer-1.x%29)
* Create a separate bug and tag it with the GTK3/GST1.x label

**Comment by genodeftest** [comment]
Are there any plans when to release Exaile 4.0?
What should be done before that?
[ ] Test and fix plugins until they are either working correctly ("The perfect") or (temporarily) removed from exaile ("The parking lot"). See [Exaile 4.x Plugin development status list](https://github.com/exaile/exaile/wiki/Exaile-4.x-Plugin-development-status-list)
[ ] fix all (or most?) bugs with [Milestone 3.4.6](https://github.com/exaile/exaile/milestones/Exaile%203.4.6). Do you plan to do a separate 3.4.6 release?
[ ] fix all (or most?) bugs with [Milestone 4.0](https://github.com/exaile/exaile/milestones/Exaile%204.0).
[ ] make sure Exaile 4.0 can be built for Windows and OS X
Am I missing something?

**Comment by virtuald** [comment]
@genodeftest there's a few key performance (freezing) problems that I'm experiencing that I want to look at before an initial beta release.

I'm undecided on a separate 3.4.6 release, I think I'd rather focus on Exaile 4 instead. I was hoping to get it out this month.

I think that the beta release can be done before all of the bugs are fixed, the things in the 4.0 milestone are things that I think should be fixed before a 4.0 final release.

I was able to successfully build a Windows installer earlier this week. I'm optimistic that I can get the OSX version working this weekend.

**Comment by Grinnz** [comment]
FYI, it is being proposed in Fedora to drop Gstreamer 0.10 due to upstream inactivity and security issues: https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/HHPHRNYQJC73ZRDHEO7E7JPR47UQABAV/#HHPHRNYQJC73ZRDHEO7E7JPR47UQABAV

**Comment by sjohannes** [comment]
Thanks for the info. There are still quite a number of bugs targetted for 4.0, but it's in a pretty good state. Rather than removing the package, it may be better to distribute Exaile's current dev version if the package maintainer is ok with a few new bugs here and there, nothing too serious.

There is a [draft release notes](https://github.com/exaile/exaile/wiki/Exaile-4.0-release-notes) that shows most of the changes in the dev version, including new dependencies and so on. There is also another page detailing [the state of all plugins included with Exaile](https://github.com/exaile/exaile/wiki/Exaile-4.x-Plugin-development-status-list).

And as always, if package maintainers need anything from our side (e.g. fixing buildfiles, checking dependencies, etc.), we're more than happy to help.

**Comment by genodeftest** [comment]
As this bug turns out to become quite general, I'd like to go back to the original report:
Exaile (on git master) is now fully ported to Gtk+ 3.x, PyGObject and GStreamer 1.x, except for some plugins (see the wiki page).
The remaining bugs (threading issues, missing menu borders on windows, …) have their own reports.

How about closing this bug report and continuing the "when to release" discussion somewhere else?

**Comment by virtuald** [comment]
This bug is no longer useful as far as I'm concerned. We'll hopefully do a release in the near future, as there are no major blockers left.
